File: Domain/ValueObjects/Order/OrderStatus.cs
namespace Domain.ValueObjects.Order;

public class OrderStatus : ValueObject
{
    public static OrderStatus Pending => new OrderStatus("Pending");
    public static OrderStatus Finished => new OrderStatus("Finished");

    public string Name { get; }

    public OrderStatus(string name)
    {
        Name = name;
    }

    private static readonly List<OrderStatus> ValidStatuses = new List<OrderStatus>
    {
        Pending,
        Finished
    };

    public static bool IsValid(string status)
    {
        return ValidStatuses.Exists(d => d.Name == status);
    }

    public override IEnumerable<object> GetEqualityComponents()
    {
        yield return Name;
    }
}

File: frontend/src/domain/valueObjects/Order/OrderStatus.ts
export default class OrderStatus {
    public static readonly PENDING = new OrderStatus('Pending');
    public static readonly FINISHED = new OrderStatus('Finished');

    private static readonly validStatuses = [
        OrderStatus.PENDING,
        OrderStatus.FINISHED
    ];

    private constructor(value: string) {
        this.value = value;
    }

    public static create(value: string) {
        if (!OrderStatus.isValid(value)) {
            throw new Error(`${value} is not a valid OrderStatus`);
        }

        return new OrderStatus(value);
    }

    public static isValid(status: string): boolean {
        return OrderStatus.validStatuses.some(validStatus => validStatus.value === status);
    }

    public value: string;
}